rust/chains/hyperlane-cosmos-modules/src/interchain_gas.rs

```diff
@@ -0,0 +1,187 @@
+use async_trait::async_trait;
+use hyperlane_core::{
+    ChainResult, ContractLocator, HyperlaneChain, HyperlaneContract, Indexer,
+    InterchainGasPaymaster, SequenceIndexer, U256,
+};
+use hyperlane_core::{ChainCommunicationError, HyperlaneDomain, HyperlaneProvider, InterchainGasPayment, LogMeta, H256, H512};
+use std::ops::RangeInclusive;
+use tracing::info;
+use cosmrs::tendermint::{
+    abci::EventAttribute,
+    hash::Algorithm,
+    Hash,
+};
+use sha256::digest;
+
+use crate::{Signer, ConnectionConf, CosmosProvider};
+
+#[derive(Debug)]
+pub struct CosmosInterchainGasPaymaster {
+    conf: ConnectionConf,
+    domain: HyperlaneDomain,
+    address: H256,
+    signer: Signer,
+    provider: Box<CosmosProvider>,
+}
+
+impl CosmosInterchainGasPaymaster {
+    pub fn new(conf: &ConnectionConf, locator: ContractLocator, signer: Signer) -> Self {
+        let provider = CosmosProvider::new(conf.clone(), locator.domain.clone(), locator.address, signer.clone());
+
+        Self {
+            conf: conf.clone(),
+            domain: locator.domain.clone(),
+            address: locator.address,
+            signer,
+            provider: Box::new(provider),
+        }
+    }
+}
+
+impl HyperlaneContract for CosmosInterchainGasPaymaster {
+    fn address(&self) -> H256 {
+        self.address.clone()
+    }
+}
+
+impl HyperlaneChain for CosmosInterchainGasPaymaster {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(CosmosProvider::new(
+            self.conf.clone(),
+            self.domain.clone(),
+            self.address(),
+            self.signer.clone(),
+        ))
+    }
+}
+
+impl InterchainGasPaymaster for CosmosInterchainGasPaymaster {}
+
+#[derive(Debug)]
+pub struct CosmosInterchainGasPaymasterIndexer {
+    provider: Box<CosmosProvider>,
+}
+
+impl CosmosInterchainGasPaymasterIndexer {
+    pub fn new(conf: &ConnectionConf, locator: ContractLocator, signer: Signer) -> Self {
+        let provider = CosmosProvider::new(conf.clone(), locator.domain.clone(), locator.address, signer);
+        Self {
+            provider: Box::new(provider),
+        }
+    }
+
+    fn parse_event(&self, attrs: Vec<EventAttribute>) -> ChainResult<InterchainGasPayment> {
+        let mut res = InterchainGasPayment {
+            message_id: H256::default(),
+            payment: U256::default(),
+            gas_amount: U256::default(),
+        };
+        for attr in attrs {
+            let key = attr.key.as_str();
+            let value = attr.value.as_str();
+
+            info!("Attr: {:?}", key);
+            match key {
+                "messageid" => {
+                    res.message_id = H256::from_slice(hex::decode(value.trim_start_matches("0x")).unwrap().as_slice())
```

Comment:
Good catch. Given that the IGP event attribute is output based on the MsgPayForGas input, it is possible that it doesn't have the 0x. I'd propose we change MsgPayForGas's MessageId to by bytes instead of string. Then, the cli tx can support both if we want and we'd have consistency in how the payment is stored in the KVStore (already a byte slice) along with outputting the event attribute consistently. I'd still like it to output with the 0x prefix to be consistent with the mailbox and ISM modules.